ui/src/app.tsx

```diff
@@ -17,25 +22,65 @@ import Overview from "./overview";
 enableMapSet();
 
 type AppProps = {
-  underlayNames: string[];
   entityName: string;
 };
 
 export default function App(props: AppProps) {
+  const dispatch = useAppDispatch();
+  const underlaysApi = useContext(UnderlaysApiContext);
+  const entitiesApi = useContext(EntitiesApiContext);
+
+  const underlaysState = useAsyncWithApi(
+    useCallback(() => {
+      return underlaysApi.listUnderlays({}).then((res) => {
+        if (!res?.underlays || res.underlays.length == 0) {
+          throw "No underlays are configured.";
+        }
+
+        Promise.all(
+          res.underlays.map((u) => {
+            if (!u.name) {
+              throw "Unnamed underlay.";
+            }
+            return entitiesApi.listEntities({ underlayName: u.name });
+          })
+        ).then((entitiesResList) => {
+          dispatch(
+            setUnderlays(
+              entitiesResList.map((entitiesRes, i) => {
+                const name = res.underlays?.[i]?.name;
+                if (!name) {
+                  throw "Unnamed underlay.";
```

Comment:
nit: this should be unreachable due to above check at L42

Reply:
Yes, but unfortunately the compiler can't confirm it so it complains without the extra check.